Full package review for v0.1.0 release

joshwlambert opened the pull request:
This PR is for the full package review of {epiparameter}. 

After this review is complete I will:
- complete the release checklist
- create a Github release 
- submit {epiparameter} to CRAN

adamkucharski commented on vignettes/extract-bias.Rmd:
````diff
@@ -0,0 +1,494 @@
+```
+
+::: {.alert .alert-primary}
+From the plots in this vignette, the bias is low and precision is high
+when extracting parameters from the gamma, lognormal and Weibull distributions 
+from both percentiles of the distribution and from median and range of a data
+set.
+
+The asymmetry of percentiles and sample size of the data does not noticeably 
+influence the bias in parameter extraction.
````
Figure 2 suggests bias for extraction from median + range can be substantial for smaller sample sizes? Which makes sense, given range has high variance as a summary statistic?

joshwlambert replied:
Can you specify which part of the plot you mean? It seems for gamma and weibull that the sample size may have a small influence on bias, but for lnorm it is the largest sample sizes that have the largest biases in the sdlog parameter.

adamkucharski commented on R/epidist_db.R:
```diff
@@ -0,0 +1,96 @@
+#' # `as_epidist()`
+#'
+#' # load influenza serial interval from database
+#' edist <- epidist_db(disease = "influenza", epi_dist = "serial_interval")
```
Note that when printing a summary of `epi_dist_db` in R, `hospitalisation_to_death` is at the top of list as parameter, but not supported yet as `epidist` object I think?

joshwlambert replied:
`<epidist>` should support `hospitalisation_to_death`. Here is an example where it seems to work fine:

``` r
library(epiparameter)
epi_dist_db <- epiparam()
epi_dist_db
#> Epiparam object
#> Number of distributions in library: 118
#> Number of diseases: 23
#> Number of delay distributions: 95
#> Number of offspring distributions: 10
#> Number of studies in library: 57
#> <Head of library>
#>       disease         epi_distribution prob_distribution
#> 1  Adenovirus        incubation_period             lnorm
#> 2 Chikungunya        incubation_period             lnorm
#> 3    COVID-19 hospitalisation_to_death           weibull
#> 4    COVID-19 hospitalisation_to_death             lnorm
#> 5    COVID-19        incubation_period              <NA>
#> 6    COVID-19        incubation_period           weibull
#> <112 more rows & 53 more cols not shown>
edist <- as_epidist(epi_dist_db[3,])
#> Using Linton et al. (2020) <10.3390/jcm9020538> PMID: 32079150. 
#> To retrieve the short citation use the 'get_citation' function
#> Numerical approximation used, results may be unreliable.
edist
#> Disease: COVID-19
#> Pathogen: SARS-CoV-2
#> Epi Distribution: hospitalisation to death
#> Study: Linton et al. (2020) <10.3390/jcm9020538> PMID: 32079150
#> Distribution: weibull
#> Parameters:
#>   shape: 1.599
#>   scale: 9.926
```

<sup>Created on 2023-05-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Please let me know if there is an issue somewhere else, and I'll make an issue and resolve.

adamkucharski reviewed (commented):
Nice work on this. Have added some comments for ease of reading in vignettes, plus a few things that would be worth tweaking to help usability and interpretation.

pratikunterwegs reviewed (changes requested):
Thanks for tagging @joshwlambert, seems like a huge effort so good job getting it this far. Some comments and suggestion below, happy to discuss more and/or work through them with you.

This is the first full package review for {epiparameter}, but there have been multiple reviews of package changes before. Looks good overall, there are a few issues which are highlighted below. 

The epidemiological parameter library is probably the most useful part of this package. I think there's bound to be some confusion between a custom `epiparam` object and the package library provided via `epiparam()` - I would suggest providing the in-built library as package data loaded using `data("X")`. I'm not convinced that the need for its own class (`epiparam`) is evident - perhaps a vignette tackling the practical aspects of using `epidist` and `epiparam` objects in a workflow might be useful to explain the usefulness of `epiparam`s. E.g. perform analysis on outbreaks of same infection in different populations, save parameter estimates for each pop as individual `epidist`s, then convert all to a single `epiparam`, and write to local file for future use or sharing (including creating `epidist`s from each row of the custom `epiparam`).

Have not gone over the files epidist_utils.R and extract_param.R, or over the vignettes in much detail as these have already been scanned by Adam; there could probably be more vignettes tackling different use cases. Happy to scan the tests and other auxiliary files for issues as well, but the current comments might already be a lot of work for now.

# General
1. Really like the speed of the tests - 2060+ tests in 5.2s is pretty nice.
2. Could somebody remind me why we need R (>= 3.6.0)?
3. Running `goodpractice::gp()` shows only code coverage issues (96% seems fine).
4. More conceptual: what is the taxonomic resolution of the vector in the vector-borne delay distribution in a `vb_epidist`? Not sure whether diseases have very different values in different vectors, and/or whether such values are reported at species resolution but might be worth checking. See e.g. zika which has multiple competent _Aedes_ vectors.

# General technical
1. Suggest updating `.lintr` file to match {packagetemplate}
2. Suggest running {styler} on R files (23 affected) and vignettes (1 affected)
3. Suggest standardising documentation when referring to package classes - some places have `<class>` and others have links \[\`epiparam\`]
4. Documentation:
	1. Use imperative function titles, and make full stops at the end of `@param`, `@return`, or other descriptions uniform
	2. Check that all functions have titles, see e.g. `is_epidist_params()`
5. Might be good to add limits to input checking on numbers and numerics, including checking for infinity
6. There are lots of uses of `anyNA` which will return TRUE even when some elements are not NA - I'm not able to fully see the effects of this, but might be something to watch out for
7. Suggest using render-readme workflow with substitution of the package name and repo in Readme.Rmd
8. I wonder whether it might be better to remove the `call. = FALSE` from `stop()` commands in internal functions - would that make them more difficult to debug later?
9. When accessing an `epidist` from `epidist_db(disease = "covid19")` - and perhaps other cases as well - I get the message "Using \<citation\> DOI", while this is already given in the print method for `epidist` under "Study". Would it be more helpful to print the rule used to select this study, e.g. "Selecting study with largest sample size"?
10. Use `writeLines()` instead of `cat()` which is only used once.
11. Check where `checkmate::assert_character()` should be replaced with `assert_string()` - e.g. vector in `vb_epidist`, pathogen in `epidist`.
12. Check usage of `1:n` and replace with `seq_len(n)` where found, e.g.`plot.epidist()`.
13. I would recommend adding braces for `if-else` statements, even when the body is a single line.
14. Standardise use of `identical()`, especially when comparing lengths etc.
15. Suggest reconsider excessive use of `unname()` on vectors; e.g. when initialising dataframes from named vectors it's possible to set `rownames(df) = NULL` later.
16. I see the use of `I()` in some places e.g. `add_ci_limits()` - why is this used?

# Functions and files

## bind_epiparam()
1. Suggest updating Details to tweak the definition of the `epiparam` class here, as I understand that `epiparam()` returns the package library, but any `epi-` objects can be bound together to form an `epiparam` class
2. Suggest use `is.list()` rather than `inherits(x, "list")`, unless there is a reason to not?
3. Suggest updating error message at the end of if-else ladder to add data.frame as a permitted input

## cbind_epiparam()
I'm not sure whether this function is necessary. I don't see it used elsewhere in the package or showcased in the vignettes. It's already possible to assign columns to an `epiparam` object as to a dataframe.

## calc_dist_params()
1. Suggest converting the Details to a numbered list.
2. I'm not very clear on what the `prob_dist_params` argument is for - there's some input checking on it, but this argument is never operated upon - only returned after assignment at the end. Might be better to initialise it as NA in that case and take away the fn argument?

## get_citation.epiparam()
Why does this function return a list rather than a character vector? Making it a character vector would bring it in line with the functionality of `get_citation.epidist()`.

## create_epidist_uncertainty()
1. Suggest wrapping body of first `if()` in `{ }` for clarity
2. Replace `checkmate::assert_character(ci_type)` with `checkmate::assert_string(ci_type)`, as it is currently possible to pass both "credible interval" and "confidence interval"

## create_epidist_metadata()
1. Suggest using `checkmate::assert_string()` for region, transmission mode, vector, and inference method if appropriate. Lots of vector borne diseases have multiple competent vectors, but perhaps delay dists are reported per species? Worth checking, see also conceptual issue above.

## create_epidist_citation()
Remove duplicated description at the top of documentation, add function title instead.

## Data dictionary
1. I see that the last update was on 11th May, is this changed manually each time? Is there an automated way to update the data dictionary?
2. Is there a line length requirement or good practice on the descriptions in this schema? Some lines are over 80 characters.

Bisaloo reviewed (commented):
Thanks for your work on this package!

I have added comments inline but will summarize key ideas here:

- `isTRUE()` & `isFALSE()` should not be the default solution when using branches and if they are necessary, it's often good to add a small comment explaining why (unless it's a common case such as `all.equal()`.
- this is a long and complex package and I'm slightly worried that:
  - it will be difficult to learn for users
  - it will be a lot of work to maintain it in the long-term and add new features
  
  In particular, it has a very large number of functions and any action that would reduce the number of functions (including internal functions but especially exported functions) would improve readability & maintainability.
  Another sign of this is the presence of copied/pasted code chunks, especially when branching to deal with different distributions. Ideally, supporting a new probability should only require to add the low-level parameter conversion functions. A rule of thumb could be that adding a new probability distribution should not create any new exported function.
  A solution could be to have a single function for all distribution and provide the distribution as an argument. This has very nice side-effects, like the fact you can automatically dispatch to the correct function without `switch()` or `if`/`else`. For example, users could now do something like:
  ```r
  epiparam() |> 
    mutate(params = convert_summary_stats(prob_distribution, mean, sd))
  ```

  If it seems absolutely impossible to reduce the number of functions / package complexity, it might make sense to consider splitting the package (which may kind of happen by handing off the database part to the collaboratory?)

joshwlambert commented on vignettes/epiparameter.Rmd:
```diff
@@ -0,0 +1,212 @@
+
+The functions `extract_param()` handles all the extraction of parameter estimates from summary statistics. The two extractions currently supported in {epiparameter} are from percentiles and from median and range.
+
+## Contributing to {epiparameter}
```
I think it would be best to get this initial release out without linking to the WHO Collaboratory, and then subsequent releases can be linked once we have adopted their database.

joshwlambert commented on vignettes/extract-bias.Rmd:
```diff
@@ -0,0 +1,494 @@
+# calculate absolute difference between true parameter and estimated value
+results <- cbind(
+  results,
+  diff_param_1 = abs(results$param_1 - results$estim_param_1),
```
Thanks for noting. I've made an issue from this comment to log it and we can consider if it might be something we want to include in the vignette after the package is released.

joshwlambert commented on R/bind_epiparam.R:
```diff
@@ -0,0 +1,119 @@
+#' The `<epiparam>` returned by `bind_epiparam()` contains the matching columns
+#' of the input objects. Therefore, if one of the input object extra columns
+#' added with are not present in the other input object these will be missing
+#' from the returned object. This also applies whether binding other
```
Resolved in 0b5e401bc1df3b79f224e775bd84d3dd0f4742bd

joshwlambert commented on R/calc_dist_params.R:
```diff
@@ -0,0 +1,157 @@
+#' `calc_dist_params()` function computes the parameters for the gamma,
+#' lognormal and Weibull distributions.
+#'
+#' This function can provide a convinient wrapper around the conversion and
```
Resolved in bc54eb4491ad9c2a6f1fac5bc233d5ceab504bdd

joshwlambert commented on R/calc_dist_params.R:
```diff
@@ -0,0 +1,157 @@
+#' lognormal and Weibull distributions.
+#'
+#' This function can provide a convinient wrapper around the conversion and
+#' extraction functions it is not known which summary statistics can be used to
```
Resolved in c712fbb2d128e286dda344c0a5a33334579804e0

joshwlambert commented on R/calc_dist_params.R:
```diff
@@ -0,0 +1,157 @@
+#' extraction functions it is not known which summary statistics can be used to
+#' calculate the parameters of a distribution.
+#'
+#' @details The hierarchy of methods are: conversion is prioritised if the mean
```
Resolved in c712fbb2d128e286dda344c0a5a33334579804e0

joshwlambert commented on R/epidist_db.R:
```diff
@@ -0,0 +1,96 @@
+#' be specified. For now this must match the author entry in the database
+#' exactly to be recognised.
+#'
+#' @details The current criteria for chosing a distribution when multiple
```
Resolved in dbde8a228c66f991c96261c173bd15c4df952608

joshwlambert commented:
@pratikunterwegs thanks for the comments and suggestions. To answer some of your questions:

> I would suggest providing the in-built library as package data loaded using data("X")

We decided to keep the data internal (see #5) and provide a reading function in order to have better control of the data that is read by the user and give the data attributes/class. Given the use of `epiparam()` and the documentation I will keep the current implementation for now.

> perhaps a vignette tackling the practical aspects of using epidist and epiparam

This vignette is planned (see #75) but will not be included in this PR and initial release.

> Could somebody remind me why we need R (>= 3.6.0)?

This version is used to gain some benefits around inheriting S3 generics (see [details in blog post](https://epiverse-trace.github.io/posts/extend-dataframes/#compatibility-with-dplyr))

> what is the taxonomic resolution of the vector in the vector-borne delay distribution in a vb_epidist? 

Currently the common name or taxonomic binomial name is accepted in the database. This is more for flexibility and to provide a vector even when the specific species is not reported. In future it may be useful to provide more resolution, but as this is a feature that is not widely used in the package at the moment I will leave things as they are.

> When accessing an epidist from epidist_db(disease = "covid19") - and perhaps other cases as well - I get the message "Using <citation> DOI", while this is already given in the print method for epidist under "Study". Would it be more helpful to print the rule used to select this study, e.g. "Selecting study with largest sample size"?

This feature follows how {socialmixr} reminds of citation information (see #97). I will add the rule to the message. 

> I wonder whether it might be better to remove the call. = FALSE from stop() commands in internal functions - would that make them more difficult to debug later?

The `call. = FALSE` is for consistency across external and internal functions. For internal functions, running them interactively allows for rerun with debug so I don't foresee the `call. = FALSE` causing any issues.

> I see the use of I() in some places e.g. add_ci_limits() - why is this used?

This was primarily used to keep the structure when adding elements to a data frame. In this case it was to add a vector to a data frame cell. There are other ways to do this (assigning with `[[` I think). Let me know if you think there is a preferable way to do this.

pratikunterwegs commented:
Thanks for explaining, I think those choices seem fine.

> >     I see the use of I() in some places e.g. add_ci_limits() - why is this used?

> This was primarily used to keep the structure when adding elements to a data frame. In this case it was to add a vector to a data frame cell. There are other ways to do this (assigning with [[ I think). Let me know if you think there is a preferable way to do this.

Regarding this, depending on how you're assigning, it could also be done by converting the vector to a list and assigning as `df$col <- list(values)`. This still allows `rbind()` with another dataframe (with the same column types? - unsure).

Might be worth explaining the use of `I()` in a comment.

joshwlambert commented on R/bind_epiparam.R:
```diff
@@ -0,0 +1,119 @@
+
+  # return epiparam
+  epiparam
+}
```
I see your point. This could be a principle of the package that we could document in a design vignette. For example, the `<epiparam>` is class is thin wrapper around `<data.frame>`, for all data manipulation use `{dplyr}`.

I think the best way to proceed to to keep the namespace similar to now for the release and have the next release target be to minimise parts of the namespace by merging or removing what we feel is unimportant.

joshwlambert replied:
Resolved in #171.

joshwlambert commented on R/bind_epiparam.R:
```diff
@@ -0,0 +1,119 @@
+    stopifnot(
+      "<data.frame> given must have the same column names as <epiparam>" =
+        setequal(colnames(epi_obj), colnames(epiparam))
+    )
```
Resolved in 243e3ca363e74405aad91db8acb17189c36b8eee

joshwlambert commented on R/create_prob_dist.R:
```diff
@@ -0,0 +1,128 @@
+                             discretise,
+                             truncation) {
+
+  if (isTRUE(discretise)) {
```
No reason, happy to remove.

joshwlambert commented on R/epidist_db.R:
```diff
@@ -0,0 +1,96 @@
+    arg = clean_disease(disease),
+    choices = clean_disease(unique(eparam$disease)),
+    several.ok = FALSE
+  )
```
Resolved in 91ad427e669a27b440856a978e4247fc93bd9773

joshwlambert commented on R/epidist_db.R:
```diff
@@ -0,0 +1,96 @@
+  )
+
+  # filter based on pathogen and delay distribution
+  eparam <- eparam[clean_disease(eparam$disease) == disease, ]
```
Resolved in 91ad427e669a27b440856a978e4247fc93bd9773

joshwlambert commented on R/epidist_utils.R:
```diff
@@ -0,0 +1,795 @@
+    return("No citation available")
+  }
+
+  # change author formatting if multiple authors or et al
+  author <- gsub(
+    pattern = "_",
+    replacement = " ",
+    x = author,
+    fixed = TRUE
+  )
+  author <- gsub(
+    pattern = "etal",
+    replacement = "et al.",
+    x = author,
+    fixed = TRUE
+  )
+
+  # check if study has two authors and if so insert ampersand
+  num_authors <- length(unlist(strsplit(x = author, split = " ", fixed = TRUE)))
+  if (identical(num_authors, 2L)) {
+    author <- gsub(pattern = " ", replacement = " & ", x = author, fixed = TRUE)
+  }
```
I think it may be better to enforce this formatting on the database side. I will make an issue noting this and then we can address it in it's own PR.

joshwlambert commented on R/epidist_utils.R:
```diff
@@ -0,0 +1,795 @@
+    function(x, y) x == y,
+    y = names(prob_dist_params)
+  )
+  is_valid_params <- any(unlist(lapply(matches, all)))
```
Resolved in db6fcfcf7536af4655e68ee3ffb46ba94a36e948

joshwlambert commented on R/epiparam.R:
```diff
@@ -0,0 +1,370 @@
+  epiparam <- new_epiparam(epi_dist = epi_dist)
+
+  # validate epiparam object
+  validate_epiparam(epiparam)
```
It shouldn't be the case. The validation step is there just in case. Would you recommend removing it?

joshwlambert commented on R/epidist_utils.R:
```diff
@@ -0,0 +1,795 @@
+                                    DOI = NA_character_) {
+  # check input
+  checkmate::assert_character(author)
+  checkmate::assert_number(year, na.ok = TRUE)
```
Can you please pinpoint where `year = NA` causes an error? I can't see a why to make this change work without a numeric default argument for `year`, but I think `NA` is preferable.

joshwlambert replied:
I've now reverted this change in 9bab77f

joshwlambert commented on R/epiparam.R:
```diff
@@ -0,0 +1,370 @@
+  logic_col <- epiparam_logic_fields(params)
+  params[logic_col] <- vapply(
+    params[logic_col], FUN = as.logical, FUN.VALUE = logical(nrow(params))
+  )
```
I think so but would be difficult to be 100% sure in this PR. I'll make an issue and check whether the JSON validation workflow covers this type checking.

joshwlambert commented on R/epidist_utils.R:
```diff
@@ -0,0 +1,795 @@
+    "Geometric must be 'prob'",
+    sep = "\n"
+  )
+}
```
Resolved in a203ee2c2f6096a4a970ea4a57cf7f73beff8381

joshwlambert commented on R/epiparam.R:
```diff
@@ -0,0 +1,370 @@
+#'
+#' @examples
+#' eparam <- epiparameter:::new_epiparam("all")
+new_epiparam <- function(epi_dist = character()) {
```
The reason for the empty types was just a convention I picked up from [Adv R](https://adv-r.hadley.nz/s3.html?q=s3#s3-constructor), but happy to remove them.

joshwlambert replied:
Resolved in 028d741f7aaeff1be894a76949ca87056b427381

joshwlambert commented on R/epiparam_utils.R:
```diff
@@ -0,0 +1,557 @@
+    } else {
+      pmid <- NA_integer_
+    }
+  }
```
This is planned, but cannot be implemented with the current information in the database as the `<bibentry>` requires the article title and journal. I will add this in a future PR, and will make an issue for now.

joshwlambert commented on R/epiparam.R:
```diff
@@ -0,0 +1,370 @@
+  )
+
+  # convert intervals from strings to numeric vectors
+  ci_cols_index <- grep(pattern = "_ci_limits$", colnames(params))
```
I've now reverted this change in baf5d6b

joshwlambert commented on R/epiparam_utils.R:
```diff
@@ -0,0 +1,557 @@
+      test = is.na(unname(params["shape"])),
+      yes = NA_real_,
+      no = params["shape"]
+    ),
```
Resolved in 7d3ded1

joshwlambert commented on R/extract_param.R:
```diff
@@ -0,0 +1,512 @@
+#' @keywords internal
+check_optim_conv <- function(optim_params_list,
+                             optim_params,
+                             optim_conv,
```
Resolved in 1538c7d0db1e35ce13d1bf3d2e6d9046b3126cb7

joshwlambert commented on R/extract_param.R:
```diff
@@ -0,0 +1,512 @@
+#' @keywords internal
+#' @author Adam Kucharski
+#' @name extraction_functions
+fit_function_lnorm_range <- function(param, val) {
+  # Median square residual
+  median_sr <- (stats::plnorm(
+    val[["median"]],
+    meanlog = param[["meanlog"]],
+    sdlog = param[["sdlog"]]
+  ) - 0.5)^2
+
+  # Probability of obtaining min, max and range:
+  min_p <- stats::dlnorm(
+    val[["lower"]],
+    meanlog = param[["meanlog"]],
+    sdlog = param[["sdlog"]]
+  )
+  max_p <- stats::dlnorm(
+    val[["upper"]],
+    meanlog = param[["meanlog"]],
+    sdlog = param[["sdlog"]]
+  )
+  range_p <- (stats::plnorm(
+    val[["upper"]],
+    meanlog = param[["meanlog"]],
+    sdlog = param[["sdlog"]]
+  ) -
+    stats::plnorm(
+      val[["lower"]],
+      meanlog = param[["meanlog"]],
+      sdlog = param[["sdlog"]]
+    ))^(val[["n"]] - 2)
+
+  # Range likelihood
+  range_sr <- -(min_p * max_p * range_p)
+
+  # Total value to be minimised
+  range_sr + median_sr
+}
```
I've implemented a working version of your suggestion. Some of the tests pass (i.e. given identical results within a tolerance), but I cannot get all of the numerical output to aline with the previous implementation. I guess this is due to the discretisation of the distribution. What do you think, did you expect the results to be exactly the same as before?

joshwlambert replied:
Resolved in c6c138a11d1dcbde868543ccda9b66e6fef4074d. Did not use the implementation suggested but similar implementation is achieved anyway.

joshwlambert commented on R/get_percentiles.R:
```diff
@@ -0,0 +1,107 @@
+#' @keywords internal
+#'
+#' @examples
+#' \dontrun{
```
Resolved in b5e49a5dd14be7cfa8b948618b36d48762a91397

joshwlambert commented on R/extract_param.R:
```diff
@@ -0,0 +1,512 @@
+}
+
+#' @rdname extraction_functions
+fit_function_lnorm <- function(param, val) {
```
Blocked based on comment above.

joshwlambert replied:
Resolved in c6c138a11d1dcbde868543ccda9b66e6fef4074d.

joshwlambert commented:
@Bisaloo thanks for the comments.

In particular the need to reduce the package API/namespace. I think this would be a good direction to go for subsequent releases, and in general I think the code base can be simplified in a number of areas. 

I like the idea of:
``` r
epiparam() |> 
  mutate(params = convert_summary_stats(prob_distribution, mean, sd))
```

But given that there is a vignette for the current conversion functions, I would need to rewrite the vignette for this change. Therefore I propose this to be a good task for v0.2.0.

Bisaloo commented:
I would strongly encourage you to do this before the first CRAN release as breaking changes are very annoying to deal with once the package is on CRAN. You'd have to either use a safe but complex deprecation process or pull the plug without safeguards and potentially cause a lot of issues for the early users, who as a result may not return to this package.

CarmenTamayo commented on README.Rmd:
````diff
@@ -0,0 +1,105 @@
+```
+
+The library is an `<epiparam>` class, which underneath is a data frame.
+
````
Reading this from a user perspective I think that mentioning these different object classes could be confusing and not really important or significant for an external user. It would be necessary to either remove the text about the classes altogether or, if it's kept, add an explanation of what these classes are and why they are converted, which I don't know if this fits within a quick start section.

I was speaking with @jamesmbaazam about design vignettes and I think that this information about the object classes would be more fitting for this type of vignette rather than the Readme

joshwlambert replied:
Thanks for raising this point. It's definitely an aspect that might confuse newer users that are unfamiliar with classes. I will reference this in an issue and make the change in an upcoming pull request.

joshwlambert commented:
## bind_epiparam()

> Suggest updating Details to tweak the definition of the `epiparam` class here, as I understand that `epiparam()` returns the package library, but any `epi-` objects can be bound together to form an `epiparam` class

The documentation states that any `epi` object can be bound to an existing `<epiparam>` object. The function does not bind any two `epi` objects.

> Suggest use `is.list()` rather than `inherits(x, "list")`, unless there is a reason to not?

`is.list()` and `inherits(x, "list")` are not functionally equivalent. Difference is due to it being atomic/recursive or inheriting from a list.

``` r
> is.list(data.frame())
[1] TRUE
> inherits(data.frame(), "list")
[1] FALSE
```

> Suggest updating error message at the end of if-else ladder to add data.frame as a permitted input

Done

joshwlambert commented:
## cbind_epiparam()

> I'm not sure whether this function is necessary. I don't see it used elsewhere in the package or showcased in the vignettes. It's already possible to assign columns to an `epiparam` object as to a dataframe.

> Would it make sense to just rely on `dplyr::add_column()` for this? I see that epiparameter is exporting a very long list of functions and I'm worried it'll be very difficult / time-consuming to maintain in the long run.

I have removed the `cbind_epiparam()` function and added a line to the getting started vignette about the use of `dplyr::bind_col()` and `tibble::add_column()`.

joshwlambert commented:
> Could the citation field in epidist objects be stored in a strict structure (e.g., `bibentry()`) rather than free text to avoid complicated and error-prone regexes here and in `create_epidist_citation()`?

Citations are handled as `<bibentry>` since PR #168.

joshwlambert commented:
## calc_dist_params()

>1. Suggest converting the Details to a numbered list.
>2. I'm not very clear on what the `prob_dist_params` argument is for - there's some input checking on it, but this argument is never operated upon - only returned after assignment at the end. Might be better to initialise it as NA in that case and take away the fn argument?

These changes as well as changes to other `<epidist>` utility functions have been made in PR #175

joshwlambert commented:
Closing this PR without merging. Initially I had started pushing to the `review` branch but subsequently decided it would be better to tackle related issues in separate, smaller pull requests (which include the changes initially made in `review`).

Thanks to all reviewers for taking the time to look through the code and make comments and suggestions. 

The PR containing the changes are linked above in this PR.

I will now work to finalise the package ready for an official v0.1.0 release.
